src/transforms/babel.js

```diff
@@ -26,19 +21,81 @@ module.exports = async function(asset) {
   }
 };
 
-async function shouldTransform(asset) {
+async function getConfig(asset) {
+  // If asset is marked as an ES6 modules, this is a second pass after dependencies are extracted.
+  // Just compile modules to CommonJS.
   if (asset.isES6Module) {
-    return true;
+    return {
+      plugins: [require('babel-plugin-transform-es2015-modules-commonjs')]
+    };
+  }
+
+  if (asset.babelConfig) {
+    return asset.babelConfig;
+  }
+
+  // Attempt to generate a config
+  let config = await getEnvConfig(asset);
+
+  // If this is the app module, and there is a .babelrc, merge it into the config.
+  // Don't load .babelrc for node_modules. https://github.com/parcel-bundler/parcel/issues/13.
+  if (!asset.name.includes('/node_modules/')) {
+    let babelrc =
+      (asset.package && asset.package.babel) ||
+      (await asset.getConfig(['.babelrc', '.babelrc.js']));
+    if (babelrc) {
+      config = Object.assign({}, babelrc, {
+        plugins: (babelrc.plugins || []).concat(config ? config.plugins : [])
+      });
+    }
+  }
+
+  return config;
+}
+
+module.exports.getConfig = getConfig;
+
+async function getEnvConfig(asset) {
+  // Load supported source engines (node and browser versions) for source module and target app,
+  // and generate a babel-preset-env config for each.
+  let sourceEngines = await getTargetEngines(asset);
+  let targetEngines = await getTargetEngines(asset, asset.options.mainFile);
+  let sourceEnv = await getEnvPlugins(sourceEngines);
+  let targetEnv = await getEnvPlugins(targetEngines);
+
+  if (!sourceEnv || !targetEnv) {
+    return;
+  }
+
+  // Do a diff of the returned plugins. We only need to process the remaining plugins to get to the app target.
+  // If this is the app module, the source and target will be the same, so just compile everything.
+  if (asset.name.includes('/node_modules/')) {
+    let sourcePlugins = new Set(sourceEnv.map(p => p[0]));
+    targetEnv = targetEnv.filter(plugin => {
+      return !sourcePlugins.has(plugin[0]);
+    });
```

Comment:
I wonder if there's any possible way to make a Babel preset handle this automatically. For example we'd like our ejected users to just have `babel-preset-react-app/dependencies` for non-app code, and have that automatically do the diffing.

Reply:
Yeah totally, I was thinking that too. The only piece of information it needs is the entry point to the bundler (`asset.options.mainFile` in above in parcel's case) in order to resolve the target app's browser list. If this information were somehow provided to babel plugins via bundlers it would be possible to do in a babel preset (or even in env itself). Maybe just resolving based on the `process.cwd` would be good enough though.

Reply:
The other thing we need in Parcel's case is to track the config files we load (e.g. `package.json`, `.browserslistrc`, etc.) so they can be added to the file watcher and cache correctly. That way when you change your browser support file, we can recompile files automatically. Would be cool if Babel presets could emit dependencies as well.

Reply:
I think I we are open for this stuff at babel, just need to specify the requirements, how this would work etc. Dunno if anyone want to focus on that at the moment as we'd like to release stable babel@7 - we could always use an outside contribution for this though 😉 

The problem is that babel at the moment has no notion of project, it operates on single files within a directory structure. And it comes with support for nested config files, overrides and stuff, which could complicate things.

But I guess what is really interesting here is a single "targets" field/config for a project, usually kept at the root of the project which is a first package.json NOT within a node_modules, it should be quite easy to locate it without much hassle and without introducing much complexity to the codebase. 

Let's open an issue on babel boards to discuss this further, we'd want to have a short summary of the problem and proposed solution for it as a discussion starter.

Reply:
>But I guess what is really interesting here is a single "targets" field/config for a project, usually kept at the root of the project which is a first package.json NOT within a node_modules, it should be quite easy to locate it without much hassle and without introducing much complexity to the codebase.

Just to clarify, I don't think the *targets* are relevant here. Env preset already can determine target browsers desired by the **app** (through browserslist).

What seems missing from Babel's side is inferring a **library's** supported `engines` and then using *those* as "minimal guaranteed" environment heuristic.

For example if my app specifies `IE 11` as a target via `browserslist`, we know ES6 needs to be compiled down. But if React says via `engines` that it runs on Node `>= 0.12` we know that it can't possibly include ES6 code. Therefore we don't have to run React through ES6 transforms. For this to work, we need presets to know they're dealing with React package (and what `engines` it has in `package.json`).

Does this make sense? Sorry if you said the same and I misunderstood.

Reply:
We are on the same page on this, I just forgot to mention need for a field in packages - hence the confusion. 

There are related discussions about similar topics on babel boards, cannot find them on mobile so will post links 2morrow.

Im not sure if it has been discussed here but we'd have to standardize a new field for package.json - smth like "source". Using "main"/"module" at the moment would cause too much problems at this point

Reply:
Why do we need that field? I'm not quite sure I'm following. At least in CRA I don't plan to compile third-party packages from source—maybe I misunderstood your intention?

Reply:
Well, I've chosen the unfortunate name for this just before hitting the bed. I didnt mean the literal "source", transpiling is becoming much more than just new shiny (often experimental) syntax with typings, macros, prevals and other cool stuff. We do not want to ship all that stuff obviously to npm.

We'd have to recommend shipping (somehow) "transpiled down to latest (?) syntax" to npm. It's a problem as latest is a moving target. What if one sets up his build tools with let's say babel@7 in place and everything works, but babel@8 gets out and "latest" is different - new syntax got standardized and people start publishing their libraries with babel@8 in mind? I believe we can figure out this problem, but we need to evaluate how it can be solved.

Many people believe that fields like "main" shouldn't require much tooling and preferably they should be runnable at least in all LTS versions of node. Using "main" for this whole idea would pretty much force consumer to use babel.

We've started some conversations about similar topics [here](https://github.com/babel/babel/issues/6328).

Reply:
Sorry, I believe I completely lost you 😥 

Let me reiterate how I view history:

* npm ecosystem used to publish things in ES5 because Node used ES5
* Client-side apps started depending on npm ecosystem and also used ES5
* Adventurous folk started using ES6 in npm packages, but since Node runs ES5, they had to Babel it down to ES5
* Node started supporting ES6
* Maintainers of popular Node packages no longer want to have a build step before publish since they mainly care about Node users and think that client-side developers can compile themselves (and some client-side developers also are fine with shipping ES6 in the first place)

And how I see the solution:

* Package authors continue publishing JavaScript (any version!) to npm as long as it's only pure valid JavaScript that runs on Node and not some JSX/Flow/stage-X. There's no separate entry point necessary for "new" or "old" code, they just publish the JS code like they always did.
* Client-side app developers compile `node_modules` using `babel-preset-env` that considers their target (e.g. "IE 9"). We already know `node_modules` always contain *only* pure JS so this always works, and code in packages never depends on specific Babel versions.
* As a potential *optimization*, we could *skip* compiling some packages down to our app's target (i.e. "iOS 10") if we know they already claim to be compatible with a "lower" target (e.g. Node 0.12). I want to reiterate this is an *optimization* and does not fundamentally change anything, except it lets us run less transformation. The output wouldn't change because we assume those features that we planned to compile don't exist in the package (per its claim of compatibility with a known lower target).

Where does my picture not match yours? I don't quite see how the `source` (or `latest` etc) entry point fits into this picture.

Reply:
In other words, what I'm trying to say is that I don't think we should solve the problem of "new Node version comes out that supports ES10, and thus popular packages released ES10 versions, making them incompatible with consumers who expect ES9". I think this is up to every specific maintainer to choose which versions of Node they want to support in their package.

To the Node consumers, there is a choice: either migrate to that version of Node, or to use the previous major version of the package, or to compile it with `preset-env` and never have to think about that.

The client-side consumers can always compile with `-env`. As an *optimization*, they can skip compiling a package if claims to only contain a code written against an old enough engine (e.g. Node 4) that we know for sure it's a subset of the version our app targets (e.g. Chrome latest).

Of course, all of this only works if we agree that packages on npm can only contain pure JS code.

Reply:
Compiling "native javascript" comes with caveats whose safety only the author of the code can truly understand - for example, what if the code depends on `(() => {}).prototype` being `undefined`, but the transpilation can't offer that? It's simply not safe to transpile code you didn't write.

Reply:
That's a great point, thanks for mentioning it! I asked in a few places but haven't considered this case.

Reply:
Even if we can't transpile perfectly in some edge cases, I think its still more valuable more of the time than not transpiling anything at all and having dependencies that don't work in an app's target engines. If a dependency is written in ES6, it won't work on IE 9 either way, but it is more likely to work with transpiling than without.

Reply:
Yet, if that dependency was transpiled prepublish, the likelihood is even higher - which is why the community best practice remains, despite a few outlier packages and the react-native ecosystem, to publish a "main" that's CJS ES5.

Reply:
@gaearon 
I have in mind pretty much exactly what you have said, regarding the history, recent practices, proposed approaches. Except for this whole "main"/"some_other_field" point.

I dont exactly disagree with you on that, I only have seen this from slightly different angle. I just often see people considering "main" as some kind of "stable" es5 version. 

You envision this being an enhancement for consumers, letting package authors to chose what they want to support etc. I thought of this (not that it is any vision of mine, and I dont agree with your approach, just had different goal in mind while we've started discussing this) more of an "enhancement" provided by package authors (with different field in the package.json) for consumers. So I have considered more opt-in situation here, whereas your solution with `main` makes it more obligatory to use tooling supporting `node_modules` transpilation (babel-preset-env). 

My original thought hasnt really contradicted yours that much. With your approach this makes it kinda "ok, lets ship whatever you want, whatever you find appropriate, we can take care of the rest". With my original approach it was more like "let's ship whatever you want as main, we cant strictly forbid that, we'll take care of the rest, BUT if you want to support older envs right out of the box please use `main` for completely transpiled version of your code and `some_other_field` as an enhancement for users who'd like to benefit from less transpilation".

Reply:
Now it clicks, thanks so much for explaining! Yeah I can see how this could be valuable although indeed the shifting baseline is a bit confusing.

Reply:
I just expect most libraries supporting at least all current LTS versions which in a sense might mean "overtranspiling" `main` for a part of the user base that would like to target newer envs.

Reply:
Is there a conclusion or a next chapter to this convo somewhere? I ran into the issue this morning and had a very similar idea. I wonder if we can borrow insight from browserify's approach of libraries [explicitly defining](https://github.com/browserify/browserify-handbook/blob/master/readme.markdown#browserifytransform-field) their transforms in package.json and including them as devDeps so consumers don't need to worry about it. I imagine you're still agreeing on a baseline target to some degree, but this could let downstream compilers opt out of certain transforms. Anyway, wanted to get thoughts on that and see if this convo has developed into a best practise somewhere. Is it recommended to transpile your node_modules directory with preset-env now, both in dev and prod?